File: examples/src/main/scala/concrete/examples/scala/SingleSectionSegmenter.scala
/**
 *
 */
package concrete.examples.scala

import edu.jhu.hlt.concrete.AnnotationMetadata
import edu.jhu.hlt.concrete.SectionSegmentation
import edu.jhu.hlt.concrete.Communication
import edu.jhu.hlt.concrete.UUID

import edu.jhu.hlt.concrete.TextSpan
import edu.jhu.hlt.concrete.Section
import edu.jhu.hlt.concrete.SectionKind

import edu.jhu.hlt.concrete.util.ConcreteUtil

/**
  * This object represents a way to apply a single `SectionSegmentation` with a single `Section` to a `Communication` object.
  * 
  * @author max
  *
  */
object SingleSectionSegmenter {
  /**
    * Return the `AnnotationMetadata` for this `SingleSectionSegmenter` (a constant value).
    */
  val Metadata: AnnotationMetadata = new AnnotationMetadata()
                                            .setConfidence(1)
                                            .setTool("SingleSectionSegmenter v2.0.0-scala")
                                            .setTimestamp(System.currentTimeMillis() / 1000)
  /**
    * Return a `SectionSegmentation` object with a single `Section`. The `Section`
    * contains a single `TextSpan` with a start at character 0 and an end at the 
    * length of the Communication's text value.
    * 
    * @param comm A `Communication` that will be used to generate the `SectionSegmentation`.
    * @return A `SectionSegmentation` as described above.
    * @since 2.0.0-SNAPSHOT
    * 
    * {{{
    * scala> val comm = new Communication().setText("hello world!")...
    * comm: edu.jhu.hlt.concrete.Communication = Communication(...)
    * 
    * scala> import concrete.examples.scala.SingleSectionSegmenter
    * import concrete.examples.scala.SingleSectionSegmenter
    * 
    * scala> val singleSectionSeg = SingleSectionSegmenter.sectionCommunication(comm)
    * singleSectionSeg: edu.jhu.hlt.concrete.SectionSegmentation = SectionSegmentation(...)
    * 
    * scala> singleSectionSeg.sectionList.get(0).textSpan
    * res9: edu.jhu.hlt.concrete.TextSpan = TextSpan(start:0, ending:12)
    * }}}
    */
  def sectionCommunication(comm : Communication) : SectionSegmentation = {
    val textSpan = new TextSpan().setStart(0).setEnding(comm.getText().length())
    val section = new Section()
      .setUuid(ConcreteUtil.generateUUID)
      .setTextSpan(textSpan)
      .setKind(SectionKind.OTHER)
      
    val ss = new SectionSegmentation().setMetadata(this.Metadata).setUuid(ConcreteUtil.generateUUID)
    ss.addToSectionList(section)
    ss
  }
}

File: util/src/main/java/concrete/tools/impl/SingleSectionSegmenter.java
/*
 * Copyright 2012-2014 Johns Hopkins University HLTCOE. All rights reserved.
 * See LICENSE in the project root directory.
 */
package concrete.tools.impl;

import java.util.ArrayList;
import java.util.List;

import concrete.tools.AnnotationDiffTool;
import concrete.tools.AnnotationException;
import edu.jhu.hlt.concrete.AnnotationMetadata;
import edu.jhu.hlt.concrete.Communication;
import edu.jhu.hlt.concrete.Section;
import edu.jhu.hlt.concrete.SectionSegmentation;
import edu.jhu.hlt.concrete.TextSpan;
import edu.jhu.hlt.concrete.util.ConcreteUUIDFactory;

/**
 * @author max
 *
 */
public class SingleSectionSegmenter implements AnnotationDiffTool<SectionSegmentation> {

  private final ConcreteUUIDFactory idFactory;
  
  public SingleSectionSegmenter() {
    this.idFactory = new ConcreteUUIDFactory();
  }
  
  /**
   * Generate a {@link SectionSegmentation} with a single {@link Section} that 
   * encompasses the entire text of the {@link Communication} object.
   * 
   * @param c - a {@link Communication} object to "section"
   * @return a {@link SectionSegmentation} that can be added to the {@link Communication}
   * @throws AnnotationException if there is no text in this {@link Communication}.
   */
  /*
   * (non-Javadoc)
   * @see concrete.tools.AnnotationDiffTool#annotateDiff(edu.jhu.hlt.concrete.Communication)
   */
  @Override
  public SectionSegmentation annotateDiff(Communication c) throws AnnotationException {
    if (!c.isSetText())
      throw new AnnotationException("This tool requires the communication to have text.");
    SectionSegmentation ss = new SectionSegmentation();
    // you can directly set the member...
    ss.metadata = this.getMetadata();
    
    // or use the "setter" pattern
    ss.setUuid(this.idFactory.getConcreteUUID());
    
    List<Section> sectionList = new ArrayList<Section>();
    Section s = new Section();
    
    // note that the "setter" pattern returns the object,
    // so you can chain together operations. see below.    
    s
      .setUuid(this.idFactory.getConcreteUUID())
      .setKind("Other");
    
    // you can also make use of constructors.
    // these contain all the required fields of the object.
    // below, the 'start' and 'end' are required, thus the 
    // ctor has 2 fields. 
    TextSpan ts = new TextSpan(0, c.getText().length());
    // as above, can use "getter"s to access fields
    // no need to "build" like protobufs - all done
    s.textSpan = ts;
    
    s.label = "The entire text of this Communication";
    sectionList.add(s);
    ss.sectionList = sectionList;
    
    return ss;
  }
  
  @Override
  public final AnnotationMetadata getMetadata() {
    AnnotationMetadata md = new AnnotationMetadata();
    md.setConfidence(1.0);
    md.setTimestamp(System.currentTimeMillis());
    md.setTool("Concrete Example tools v1.0");
    return md;
  }
}
